typing/typeclass.ml

```diff
@@ -1570,6 +1574,21 @@ let final_decl env define_class
     raise(Error(cl.pci_loc, env, Non_collapsable_conjunction (id, clty, trace)))
   end;
 
+  (* make the dummy method disappear *)
+  begin
+    let self_type = Ctype.self_type clty.cty_type in
+    let methods, _ =
+      Ctype.flatten_fields
+        (Ctype.object_fields (Ctype.expand_head env self_type))
+    in
+    List.iter (fun (lab,kind,_) ->
+      if lab = dummy_method then
+        match Btype.field_kind_repr kind with
+          Fvar r -> Btype.set_kind r Fabsent
+        | _ -> ()
+    ) methods
+  end;
+
```

Comment:
You should explain somewhere the new invariant:
* the `self_type` of a final object type contain no dummy method from the beginning
* for class definitions, a dummy method is added during the definition, but removed by `final_decl`, so that immediate objects one cannot inherit from not yet finished classes.

By the way, I wonder how it could happen: it seems that one cannot refer to a not yet finished declaration anyway (`Unbound_class_2`).

Reply:
Fair point, I'll add a comment above `class_structure`.

It can happen because we can unify the self type of a "final" object with the self type of a yet unfinished class; this was previously allowed but resulted in an assert false later on (I believe this bug has been here for ever, I reproduced the assert false from 4.03 all the way to trunk).
Such an example is added in commit c3027c1 .